cartographer_ros/cartographer_ros/offline_node_main.cc

```diff
@@ -257,6 +257,17 @@ void Run(const std::vector<string>& bag_filenames) {
     }
 
     bag.close();
+
+    if (FLAGS_keep_running) {
```

Comment:
What does this help? No more data is fed into Cartographer, so it will never do the final optimization, so your map will never converge to the final result. And in the multi trajectory case you will just spin here forever and never start on the second trajectory.

I think you have to add hooks into the MapBuilder and maybe the sparse pose graph/constraint builder to give you an update on the state while it is optimizing.